bpf:ipsec: extend bpf tests for ipsec_maybe_redirect_to_encrypt

smagnani96 opened the pull request:
Being hugely inspired by https://github.com/cilium/cilium/commit/b824305182d9ce13adcd28cd94ad287f3e38178a, especially for infrastructure and subset of tests to run. Had to refactor a lot of stuff though, and tried to modify the infra so that we just have 2 files, `native` and `tunnel`. Inside each of them, they now test both IPv4 and IPv6.

In brief, this patch covers tests for `ipsec_maybe_redirect_to_encrypt()` with:
1. Native IPv4 packet
2. Native IPv6 packet
3. VXLANv4-InnerIPv4 packet
4. VXLANv4-InnerIPv6 packet

In addition, this patch lands the infra for the following tests, which cannot be enabled yet (MBOI for/after https://github.com/cilium/cilium/pull/39620):
1. VXLANv6-InnerIPV4
2. VXLANv6-InnerIPV6

smagnani96 committed:
bpf:tests: add pktgen helpers for VXLANv6 underlay and VXLANv4-InnerV6

This land basic helpers to push:

1. IPv6 underlay VXLAN: despite not being yet fully supported, this commit
   adds helpers that can be then used for testing when landing support.
2. IPv6 inner packets rather than manually having to do it step-by-step

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: VXLAN helpers support for ipsec_maybe_redirect_to_encrypt

This commit lands support to our bpf unit test suite for subsequent
commits in which we need to generate either native and VXLAN packets
for checking the ipsec egress hook function.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: adjust native IPv4 tests for ipsec_maybe_redirect_to_encrypt

This takes the current test for calling ipsec_maybe_redirect_to_encrypt
with a native packet into inline function, which accept a parameter to
determine whether the packet is ipv4 or ipv6. This is useful to introduce
IPv6 support in subsequent commit.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: add native IPv6 tests for ipsec_maybe_redirect_to_encrypt

Given the refactor in the previous commit, let's now enable also
ipsec_maybe_redirect_to_encrypt tests for native IPv6 packets.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: native IPv4 bad ids for ipsec_maybe_redirect_to_encrypt

This adds a separate IPv4 test case when calling ipsec_maybe_redirect_to_encrypt
with a native packet to make sure that, given wrong identities, the hook
does not encrypt the packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: add native IPv6 bad ids for ipsec_maybe_redirect_to_encrypt

Similar as in previous commit, let's now also enable bad identities tests
for native IPv6 packets when calling ipsec_maybe_redirect_to_encrypt, to
make sure packets don't get encrypted.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: VXLANv4-InnerIPv4 tests for ipsec_maybe_redirect_to_encrypt

Similar as in previous commits, let's now add additional tests for the
ipsec_maybe_redirect_to_encrypt hook when being called with an overlay
IPv4 packet containing an inner IPv4 packet. Given it is overlay, we now
expect it to always be encrypted, expect when already encrypted and marked
with the MARK_MAGIC_OVERLAY_ENCRYPTED mark.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: VXLANv4-InnerIPv6 tests for ipsec_maybe_redirect_to_encrypt

Complementary to the previous commit, let's now also add tests when having
and IPv4 overlay packet with an inner IPv6 packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

julianwiedmann commented on bpf/tests/ipsec_redirect_tunnel.c:
```diff
@@ -0,0 +1,191 @@
+		key.ip6 = *(union v6addr *)DST_NODE_IP_6;
+	}
+
+	map_update_elem(&cilium_node_map_v2, &key, &val, BPF_ANY);
```
This could be `node_add_entry()` instead.

smagnani96 replied:
Cool, updated accordingly, ty!

julianwiedmann commented on bpf/tests/ipsec_redirect_tunnel.c:
```diff
@@ -0,0 +1,191 @@
+
+	if (!outer_ipv4)
+		test_fatal("IPv6 Overlay is not yet supported.");
+
```
Paul just merged the datapath support for this :)

smagnani96 replied:
Amazing, last time I checked the PR wasn't merged yet.
Added last 2 commits to enable these VXLANv6-InnerV4 and  VXLANv6-InnerV6 cases.

julianwiedmann commented on bpf/tests/ipsec_redirect_tunnel.c:
```diff
@@ -0,0 +1,191 @@
+	 * CTX_ACT_REDIRECT is set.
+	 */
+	TEST("overlay-host-to-pod", {
+		ctx->mark = MARK_MAGIC_OVERLAY | (HOST_ID << 16);
```
This could be `set_identity_mark(ctx, MARK_MAGIC_OVERLAY, HOST_ID)` instead.

(same below)

smagnani96 replied:
Changed accordingly, thanks 🙏🏼

julianwiedmann commented on bpf/tests/ipsec_redirect_tunnel.c:
```diff
@@ -0,0 +1,191 @@
+		assert(ctx->mark == TARGET_MARK);
+		assert(ret == CTX_ACT_REDIRECT);
+	})
+
```
To be sure - is this a real scenario? I'm surprised we would be sending out overlay traffic that contains traffic from `WORLD_ID` ... but hmm, maybe `DSR-Geneve`?

smagnani96 replied:
I think it's possible as for the DSR-Geneve case, but I'm not sure in that case we would hit such codepath 🤔 
With MARK_MAGIC_OVERLAY set, we never check the carried identity. The aim of the test was to make sure that it would pass independently with CTX_ACT_REDIRECT.
I've changed it accordingly to test just 1 case (pod SOURCE_IDENTITY), while leaving the exploration of whether the type of traffic you're referring to for another PR/future work.

Any feedback is appreciated, and in case feel free to solve this conversation 😃

julianwiedmann replied:
Ack, let's start with this and expand as needed :)

julianwiedmann commented on bpf/tests/ipsec_redirect_tunnel.c:
```diff
@@ -0,0 +1,191 @@
+		assert(ctx->mark == TARGET_MARK);
+		assert(ret == CTX_ACT_REDIRECT);
+	})
+
```
Same here. I suppose there's no harm in testing for it, but would be cool if we could add a comment what type of traffic this would be (that's originating from a remote-node, but is then overlay-encapsulated by the local node).

smagnani96 replied:
Same reply as above.

julianwiedmann commented on bpf/tests/ipsec_redirect_native.c:
```diff
@@ -76,3 +65,141 @@ int ipsec_redirect_check(__maybe_unused struct __ctx_buff *ctx)
+	else
+		ipcache_v6_add_entry((const union v6addr *)SOURCE_IP_6, 0,
+				     SOURCE_IDENTITY, 0, BAD_SPI);
+
```
These would only become relevant if we pass `UNKNOWN_ID` to `ipsec_maybe_redirect_to_encrypt()`, correct?

smagnani96 replied:
Right, I just dropped this check 😅 
Testing with UnknownID would require this piece of code, but it wouldn't provide much value IMHO.

julianwiedmann commented on bpf/tests/ipsec_redirect_generic.h:
```diff
@@ -27,11 +27,78 @@ int mock_ctx_redirect(const struct __sk_buff *ctx __maybe_unused,
 #define SOURCE_IP v4_pod_one
-#define SOURCE_IDENTITY 0xAB
+#define SOURCE_IP_6 v6_pod_one
+#define SOURCE_IDENTITY 0xAB0000
```
(out-of-scope) I've always wondered how robust it is to pick such "random" numbers as identities. Or how easy it would be to have accidental conflicts with range-based checks like the CIDR-range check in `identity_is_cluster()`.

smagnani96 replied:
Agree, that might not be super clear.
For this suite, I've now changed and used `#define SOURCE_IDENTITY CIDR_IDENTITY_RANGE_START - 1` and `#define DST_IDENTITY CIDR_IDENTITY_RANGE_START - 2`: these values should always pass all checks.
Does this worsen readability?

julianwiedmann replied:
I think it's a great start! For the future we'll probably want to consolidate & manage these identities in a common header file (`POD_IDENTITY_1`, `POD_IDENTITY_2`, ...), similar as we do for IPs.

julianwiedmann reviewed (commented):
:wave: just a few comments inline, since this area is dear to my heart :). Feel free to decline / defer to a follow-up PR.

smagnani96 committed:
bpf:tests: VXLANv6-InnerIPv4 tests for ipsec_maybe_redirect_to_encrypt

Complementary to the previous commit, let's now also add tests when having
and IPv6 overlay packet with an inner IPv4 packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
bpf:tests: VXLANv6-InnerIPv6 tests for ipsec_maybe_redirect_to_encrypt

Complementary to the previous commit, let's now also add tests when having
and IPv6 overlay packet with an inner IPv6 packet.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
